File: 05_AWB/matlab/gray_world.m
clc;close all;clear;

%% 读取文件

img = imread("../images/NikonD5200_0001_G_AS.png");
[height,width,ch] = size(img);

r = img(:,:,1);
g = img(:,:,2);
b = img(:,:,3);


r_mean = mean(mean(r));
g_mean = mean(mean(g));
b_mean = mean(mean(b));

%% 计算gain值
% 我们都是向g_mean靠

r_gain = g_mean/r_mean;
b_gain = g_mean/b_mean;

%% 还原
new_img = zeros(size(img));
new_img(:,:,1) = r*r_gain;
new_img(:,:,2) = g;
new_img(:,:,3) = b*b_gain;

new_img(new_img>255) = 255;
new_img = uint8(new_img);

%%  展示
figure;

subplot(1,2,1);
imshow(img);
title("org");

subplot(1,2,2);
imshow(new_img);
title("gw");

File: 05_AWB/python/gray_world.py
import cv2
import numpy as np
import matplotlib.pyplot as plt

if __name__ == '__main__':
    raw_img = cv2.imread("../images/NikonD5200_0001_G_AS.png")

    img_rgb = cv2.cvtColor(raw_img,cv2.COLOR_BGR2RGB)
    r_channel = img_rgb[:,:,0]
    g_channel = img_rgb[:,:,1]
    b_channel = img_rgb[:,:,2]

    r_channel_mean = np.mean(r_channel)
    g_channel_mean = np.mean(g_channel)
    b_channel_mean = np.mean(b_channel)

    r_channel_gain = g_channel_mean/r_channel_mean
    b_channel_gain = g_channel_mean/b_channel_mean

    new_img = np.zeros(raw_img.shape)
    new_img[:,:,0] = r_channel*r_channel_gain
    new_img[:,:,1] = g_channel
    new_img[:,:,2] = b_channel*b_channel_gain

    new_img[new_img>255] = 255
    new_img = new_img.astype(np.uint8)

    plt.figure()
    plt.subplot(121)
    plt.imshow(img_rgb)
    plt.subplot(122)
    plt.imshow(new_img)
    plt.show()